test: Check complete ehrQL dataset definitions in docs

StevenMaude opened the pull request:
# Summary

This is an alternative to #1475 which separated out code examples as a prelude to testing them.
Rather than separating out all the included code examples, we test:

* complete dataset definitions — included as `ehrql` code blocks in the ehrQL documentation Markdown source under the `docs/` directory
* dataset definitions in `.py` modules anywhere under the `docs/` directory

It partially addresses #1325 by testing the complete dataset definition examples.

As a result of this testing, most of the existing examples have been validated as working in this PR, and a few corrected in #1677.

The alternative would be to use an already existing package to do this. I [summarised several of these in #1475](https://github.com/opensafely-core/ehrql/pull/1475#issuecomment-1725419508). Unfortunately, most of them don't really do exactly what we want. So having our own approach, without adding extra dependencies, seems reasonable. **We could re-review these options in future and see if any do become more suitable.**

# What gets tested

## Code blocks labelled as `ehrql`

It tests ehrQL example code blocks that are marked with the `ehrql` syntax. This is a simple example:

``````
```ehrql
…
```
``````

These code blocks still retain Python syntax highlighting. The only difference is that they get extracted and tested by the test suite.

**This test code uses the extension that extracts these code blocks out. Therefore, if a code block displays correctly in the ehrQL documentation, it should also be tested.** Allowed code block syntax in our current documentation configuration goes beyond "a line starting with three backticks"; see the [SuperFences documentation](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/).

## Python source dataset definitions

Full dataset definitions in the `docs/` directory are tested.

:warning: It is currently assumed that any `*.py` file in the `docs/` directory is a working dataset definition for now. We could establish a filename convention or comment marker in the file to indicate it should be tested.

## What isn't tested

There are other code-related examples included in the ehrQL documentation that are not checked:

* Sections or snippets of dataset definitions that are incomplete
* The contents of datasets output from dataset definitions
* Dataset definitions that deliberately fail
* The tracebacks of failing dataset definitions 
* Python interactive console sessions
* OpenSAFELY projects
* Commands given at the terminal

## How search and testing of examples works

Checking is rudimentary and works as follows for inline dataset definition examples:

1. The test code extracts code blocks from any `*.md` files in `docs/`.
1. The ehrQL `generate_dataset` function is run against the code blocks. This uses random dummy data.
1. We check there's some valid output CSV data that we can iterate through, without checking what the data is.

Broadly, the same check takes place for the Python dataset definitions:

1. The test code searches the documentation directory for `*.py` files.
1. The ehrQL `generate_dataset` function is run against the code from each file. This uses random dummy data.
1. We check there's some valid output CSV data that we can iterate through, without checking what the data is.

## Workarounds that are used

* `codelist_from_csv()` gets monkeypatched so we just return some arbitrary codelist codes, and the validation of codes is patched out. This could be improved by establishing a convention of column names in the documentation examples that are use to detect which test CSV codes should be used: for example, `codelist_from_csv(filename="my_file", column="snomedct")`

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,185 @@
+
+
+def create_example_test_case_id(example: DatasetDefinitionExample) -> str:
+    """Generate helpful test case IDs for pytest from a DatasetDefinitionExample."""
```
```suggestion
    """Create helpful test case IDs for pytest from a DatasetDefinitionExample."""
```

StevenMaude commented on tests/docs/test_complete_examples.py:
````diff
@@ -0,0 +1,185 @@
+    # Note that we're currently not excluding any paths like the generated docs.
+    # Do we need to? I don't think so, because if examples aren't labelled as
+    # ```ehrql then they won't be checked.
+    markdown_paths = list(docs_path.glob("**/*.md"))
+    return markdown_paths
````
```suggestion
    return list(docs_path.glob("**/*.md"))
```

StevenMaude commented on tests/docs/test_complete_examples.py:
````diff
@@ -0,0 +1,185 @@
+
+    for line_number, line in enumerate(file, start=1):
+        # Check if the line marks the start of a ehrQL code block?
+        if line.startswith("```ehrql") or line.startswith("``` ehrql"):
````
thought: I'd started with just trying to get the content of interest out. We could instead pull out *all* code blocks using the backtick formatting, and then filter those for the syntax(es) we want at the end. That might be `ehrql-dataset` and `ehrql-measure`, potentially.

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,185 @@
+            if in_ehrql_block:
+                # TODO: Should this fail completely?
+                # NB: This drops the incomplete block entirely.
+                warnings.warn(
+                    f"In {example.source_filename}, a new ehrQL code block starts on line {line_number} before the previous one has been closed."
+                )
```
thought: This is only checking this condition for `ehrql` code blocks. It's not going to warn if a different kind of code block is not closed:

``````
```python
some code
```ehrql
more code
```
``````

But the important context here is that this means that an example we might expect to check will not be checked, because it's incomplete, so maybe that's not too important.

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,185 @@
+# Do we need to handle measures?
+def test_ehrql_dataset_example(example: DatasetDefinitionExample) -> None:
+    """Check that an ehrQL generate dataset example works."""
+    # We need to have actual file paths to pass to generate_dataset.
```
context: This is because the dataset definition loading uses [`importlib.util.spec_from_file_location`](https://docs.python.org/3/library/importlib.html#importlib.util.spec_from_file_location), which takes a file path and then loads it, so we can't pass in a `StringIO` or similar.

Maybe there's a better way?

StevenMaude commented on tests/docs/test_find_docs_examples.py:
````diff
@@ -0,0 +1,82 @@
+
+def test_find_docs_examples_warning_end_of_file():
+    example = StringIO("```ehrql")
+    example.name = "test_find_docs_examples_warning_open_block"
````
```suggestion
    example.name = "test_find_docs_examples_warning_end_of_file"
```

StevenMaude commented on tests/docs/test_find_docs_examples.py:
````diff
@@ -0,0 +1,82 @@
+
+def test_find_docs_examples_warning_end_of_file_with_code():
+    example = StringIO("```ehrql\nsome code\n```\n```ehrql")
+    example.name = "test_find_docs_examples_warning_open_block"
````
```suggestion
    example.name = "test_find_docs_examples_warning_end_of_file_with_code"
```

StevenMaude commented on tests/docs/test_find_docs_examples.py:
````diff
@@ -0,0 +1,82 @@
+def test_find_docs_examples_working():
+    example = StringIO("```ehrql\nsome code\n```")
+    # Unlike file objects, StringIO objects do not have a name.
+    # We use this in the code being tested, to save somewhat redundantly passing the name.
````
```suggestion
    # We access the file's name in the relevant code being tested, to save somewhat redundantly passing the name.
```

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,185 @@
+    """Store details of a dataset definition example."""
+
+    source_filename: str
+    start: int = None
+    end: int = None
```
```suggestion
    start_line: int = None
    end_line: int = None
```

StevenMaude commented:
question: I was going to check the newly added dataset definition under "writing-a-more-complex-dataset-definition". However, on looking, I found [this is actually a separate snippet-included file](https://github.com/opensafely-core/ehrql/blob/236f9d72a3ed95640b205370ea6d442cc924b9ec/docs/tutorial/writing-a-more-complex-dataset-definition/index.md#L8) and *not* written inline.

So we'd need an extra test here to also run against any Python modules that are dataset definitions, and some convention (filename?) of specifying unambiguously that a Python file is a dataset definition to be tested.

What are the actual requirements here? Do we need to handle:

* only inline written dataset definitions in code blocks, as this PR does?
* only file-included dataset definitions as in #1475?
* both?

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,186 @@
+
+        temp_dataset_path = temp_directory_path / (temp_filename_base + ".csv")
+
+        # TODO: can we monkeypatch codelist_from_csv so that we can test documentation examples
+        # that use a codelist?
```
thought: This might work, I think, for the simple examples.

StevenMaude replied:
I'm having a go at this anyway, because it might not be too much work.

StevenMaude commented on tests/docs/__init__.py:
question: Is this a sensible path for these tests to live?

Or some other path?

There is a `tests/unit/docs`, but these don't feel like unit tests, they are end-to-end tests of the examples.

StevenMaude replied:
And should there be a `just` command to specifically check the examples?

StevenMaude commented:
Almost all the dataset definitions in the examples page are now tested. This is via some slightly dubious hacks in the test code.

iaindillingham reviewed (commented):
It's great to see that you've improved the examples in response to the test. I've left several comments, but in general I'd strongly suggest that for each `TODO`, you make and document a decision, and then remove the `TODO`. If you don't, then who will?

I'm uneasy about monkeypatching `codelist_from_csv`. Would it be possible to point each call to a codelist? (We'd only need one, and it need not be a real one.)

As an aside, I stepped through the PR commit by commit, but found the narrative hard to follow, as some later commits refactor some earlier commits. Rebasing can be a massive pain, but it makes for an easier review. (Dave is a master of the narrative PR.)

inglesp reviewed (commented):
It's really good to see all the errors this has found!  I'd love to see those commits pulled out of this PR and committed as soon as possible.

A few things I'd like to discuss:

* Can we simplify the markdown parsing code?  For instance, you'll be able to pull out all `ehrql` blocks with a regex.  It might not catch all the errors that you're checking for, but we'd expect mkdocs to do that.  Can we even hook into mkdoc's markdown parser?
* Should datasets always define a population?  It would simplify our code, and wouldn't add too much to the examples.
* Can we avoid mocking by providing real CSV files?  I'm a bit allergic to mocking, perhaps irrationally, but a comment on a commit makes me think that there's a risk they're going to hide errors here.
* There's quite a lot of TODOs in the code.  I've tried to give my answers to some of them.  I'd like them to be resolved before this PR's merged though.

(My review has crossed with Iain's and I've not had the chance to review his review, so we might be saying the same thing twice -- sorry about that.)

StevenMaude commented:
Thanks both for the review; the suggestions are good improvements.

StevenMaude commented:
@iaindillingham @inglesp:

* TODOs; yes, I'll make documented decisions on these and remove them.
* > Can we avoid mocking… (of `codelist_from_csv`)?
   * I'll have another go at this to see what's possible. I'm not sure if there are issues with mismatches of the types of expected codes that the function returns, and that the dataset definitions use. Likewise with the `column` argument; if different dataset definitions could be using different column names, I'm not quite sure how that will work.

@inglesp:

* > Should datasets always define a population?
  * Yes, Iain [suggested that the examples should.](https://github.com/opensafely-core/ehrql/pull/1648#discussion_r1368896038) — which takes some of the current "magic" out from the test.
* > Can we simplify the markdown parsing code? (to use either a regex or Python Markdown) … It might not catch all the errors that you're checking for, but we'd expect mkdocs to do that. 
  * I'll have another look at this. I don't believe MkDocs catches incomplete code blocks. At least, it doesn't even warn about them. But maybe Python Markdown — which MkDocs uses — can?

StevenMaude commented:
The fixes to examples are in #1677, so I'll rebasing this PR to remove those commits.

StevenMaude commented:
I looked into offloading the code block extraction to the same Markdown tooling that the documentation uses, and [it's only possible if you drop the line numbers](https://github.com/facelessuser/pymdown-extensions/discussions/2214).

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,255 @@
+            yield from find_complete_ehrql_examples_in_markdown(f)
+
+    # TODO: establish a naming convention to specify dataset definition should be tested.
+    dataset_definition_source_paths = discover_paths("**/*.py")
```
Actually, I think it's fine to leave it as is.

StevenMaude commented:
I've also added `define_population()` lines to the examples that are being tested and removed the workaround that patched this in for dataset definitions that didn't have it.

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,220 @@
+    def wrapped_codelist_from_csv(*args, **kwargs):
+        """Returns the result from ehrql.codelist_from_csv.
+
+        This is used to monkeypatch the real ehrql.codelist_from_csv
+        so that we can use a mock CSV, but it:
+
+        * validates the function arguments
+        * calls the real function with the mock CSV data
+
+        Different documentation examples may refer to different CSV columns.
+        Because of this, we change the arguments passed to codelist_from_csv().
+        """
+        assert (len(args) == 1) or kwargs.get(
+            "filename"
+        ) is not None, "codelist_from_csv arguments not set correctly"
+
+        try:
+            kwargs["column"] = "code"
+        except KeyError:
+            assert False, "codelist_from_csv arguments not set correctly"
+        try:
+            kwargs["category_column"] = "category"
+        except KeyError:
+            assert False, "codelist_from_csv arguments not set correctly"
+
+        return codelist_fn(filename=tmp_codelist_path, **kwargs)
+
+    with (
+        unittest.mock.patch(
+            "ehrql.codelist_from_csv",
+            wraps=wrapped_codelist_from_csv,
+        ),
+        # There is no codelist code that satisfies constraints of all code systems,
+        # so patch out the validity check and just pass in a fake codelist.
+        # If we adopted a standard for example column names, we could use valid codes.
+        unittest.mock.patch(
+            "ehrql.codes.BaseCode.__post_init__",
+            return_value=None,
+        ),
+    ):
```
@inglesp: I've had another go at running `codelist_from_csv()` with less mocking, so the real `codelist_with_csv()` function is now at least called, albeit with patched arguments.

(I can do more to validate the filename is a correct argument.)

We always need to patch out the filename to point a temporary CSV. If we decided on a convention for these in examples: like `column="snomedct"`, we could actually pass through the `column` and `category_column`.

This would then allow actual codes to be used: we could create the temporary CSV could have these columns.

I'm not sure if this is getting too complex to be worthwhile though.

StevenMaude replied:
(I fixed the checks here to use `inspect.signature`.)

StevenMaude commented on tests/docs/test_complete_examples.py:
````diff
@@ -0,0 +1,226 @@
+
+    Incomplete ehrQL dataset definitions should be labelled as ```python,
+    and not with ```ehrql."""
+    # There are three kinds of possible ehrql code blocks that we might see.
````
[This isn't true](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#nested-fence-format), I'll have a go at correcting this.

StevenMaude commented:
Because it was bugging me, I've had a go at trying out different routes to extracting the code blocks:

* The existing very naive approach which is simple, but could miss blocks:
  ````
  ```ehrql
  …
  ```
  ````
* Using the SuperFences regexes to match more flexibly as to what code blocks actually allowed in the documentation, which is more tricky to get right without introducing potential bugs or inconsistencies.
* Using the custom fences feature of SuperFences to extract the blocks.
  * This does have the advantage of not writing our own code to extract the blocks, and, because we use the same code, we know that we will extract the same blocks that the documentation "sees" when being built.
  * The downsides are: no line numbers — only the number of the block in the page, so we'd have to print any failing code instead to help identify it — and without direct warning of missing the closing blocks. (But indirectly running the code should still catch failures, because code in an unclosed example instead would get amalgamated with the next example and intervening text. Except for examples at the end of the page, but these would be complete, just unclosed, and would just be a formatting issue, not a code issue.)
  * So I'm leaning towards this right now.

Can tidy and write up a bit more next week.

StevenMaude commented:
On my computer, this used to take about 10-12 seconds to run these tests, and now takes over 60 seconds following the sandbox PR #1653.

It also breaks my monkeypatching of `codelist_from_csv()` so a workaround is needed.

evansd commented:
@StevenMaude I'd suggest not using `main.generate_dataset` to test these but rather using `loaders.load_dataset_definition_unsafe()`:
https://github.com/opensafely-core/ehrql/blob/eae0ece1e197bd089e46a51e45e0e40b20f94409/ehrql/loaders.py#L229

This will be much faster and allow your monkeypatching to work.

StevenMaude commented:
@evansd: I'd hope to get rid of the monkeypatching and maybe use real CSVs in future (#1694); but I'll bypass it for now; thanks.

StevenMaude commented:
# What's changed from when this PR was last reviewed:

* We now use the [SuperFences extension](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/) to extract the Markdown fences. This is the extension we're actually using when building the documentation to do so. This approach has several advantages, and a couple of disadvantages over the previous very crude regex —  matching ` ```ehrql ` at the start of a line — approach 
  * :heavy_plus_sign: Internally, it's still using regular expressions for matching, but they are abstracted away. We're also using an [approach that the pymdownx developer approves of](https://github.com/facelessuser/pymdown-extensions/discussions/1291).
  * :heavy_plus_sign: It is clear when previewing the documentation what will be tested: anything with an `ehrql` syntax that is displayed as a code block.
  * :heavy_plus_sign: Further, there should be no risk of including code blocks that have a valid appearance but are not tested.
  * :heavy_plus_sign: Though not implemented, we could in future repurpose the [attribute features of fences](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#injecting-classes-ids-and-attributes) for more. (These are also in plain Python Markdown code fences, but our simple regex approach didn't pull these out.)   
  * :heavy_minus_sign: We lose the warnings for incomplete code blocks that were previously present. This is mitigated by:
    * incomplete fences at the end of a Markdown source file not being formatted as a code block, so this should eventually get spotted
    * fences that are accidentally left open but closed by a subsequent fence will include something like: two dataset definitions and intervening text, and most likely fail to run 
  * :heavy_minus_sign: We lose the direct line number referencing for Markdown fences. It is [not possible to extract these](https://github.com/facelessuser/pymdown-extensions/discussions/2214#discussioncomment-7394383). In place, I've extracted the ordinal position as "fence number", and, on error, the whole dataset definition is displayed.
 
# What I'm happy to defer

* #1694
* #1697 

These are self-contained improvements that could be made on top of this PR. This PR has been open long enough that it's probably preferable to merge it soon, to get the benefit of continuously testing some of the examples. Both of the issues describe potential self-contained improvements.

StevenMaude commented:
Failure is due to Cloudflare's current API glitches as of 2023-11-02:

> `Failed to get Pages project`

StevenMaude commented on tests/docs/test_run_generate_dataset_example.py:
```diff
@@ -0,0 +1,92 @@
+        )
+
+
+def test_run_generate_dataset_example_gives_unreadable_csv(tmp_path):
```
thought: There are potentially more tests we could add here; for example, checking that ` ```{.ehrql} ` works, but as we're not using that feature yet — maybe we would if wanting to add extra inline data — I haven't.

StevenMaude commented on mkdocs.yml:
```diff
@@ -71,8 +71,10 @@ markdown_extensions:
   - toc:
       permalink: "🔗"
-  - pymdownx.highlight
-  - pymdownx.superfences
```
note: We already included `pymdownx.superfences` in the lines above (that are unchanged, and not displayed by default).

StevenMaude commented on tests/docs/test_run_generate_dataset_example.py:
```diff
@@ -0,0 +1,91 @@
+    )
+    with pytest.raises(test_complete_examples.DatasetDefinitionTestError) as exc_info:
+        test_complete_examples.test_ehrql_generate_dataset_example(tmp_path, example)
+    assert type(exc_info.value) is test_complete_examples.DatasetDefinitionTestError
```
comment: not currently in the Pytest docs at time of writing, but [mentioned in this PR](https://github.com/pytest-dev/pytest/pull/11578)

StevenMaude commented on tests/docs/test_run_generate_dataset_example.py:
```diff
@@ -0,0 +1,91 @@
+            """
+        ),
+    )
+    with unittest.mock.patch("ehrql.main.generate_dataset", return_value=None):
+        with pytest.raises(
+            test_complete_examples.DatasetDefinitionTestError,
+            match=r"Check of output dataset CSV failed for example",
```
question: I actually wanted to use [exception notes as added to Python 3.11](https://docs.python.org/3/tutorial/errors.html) in place of a custom exception to add these messages as notes to the original exception.

pytest [apparently supports these](https://github.com/pytest-dev/pytest/issues/11223) including using `match` to check against the notes. But I was finding that pytest was only seeing the first line of a `FileNotFoundError` exception which I was trying to raise.

```
AssertionError: Regex pattern did not match.                                                                                                                                                                 
 Regex: 'Check of output dataset CSV failed for example'                                                                                                                                                     
 Input: "[Errno 2] No such file or directory: /tmp/… 
```

I don't know if I'm doing something wrong in how I was trying to raise the exception. In [this code](https://github.com/opensafely-core/ehrql/pull/1648/commits/e4b718eb4a683b31b2007329e86c3c4e18a4beb3), I was doing:

```python
try:
    validate_dataset_output(tmp_dataset_path)
except Exception as e:
    e.add_note(f"Check of output dataset CSV failed for example: \n{formatted_example}")
    raise
```

If anyone think exception notes are a better approach and knows what I'm doing wrong, I'd love to know!

StevenMaude commented on tests/docs/test_complete_examples.py:
```diff
@@ -0,0 +1,279 @@
+        self.fences.append(MarkdownFence(source=src, language=language))
+        return ""
+
+    def _configure_superfences(self):
+        """Retrieves the existing extensions settings from the mkdocs.yml
+        configuration, replacing any custom SuperFences fences with a special
+        test custom fence to extract all fences."""
+        config_path = Path(__file__).parents[2] / "mkdocs.yml"
+        config = mkdocs.config.load_config(config_file_path=str(config_path))
+        assert "pymdownx.superfences" in config["markdown_extensions"]
+        config["mdx_configs"]["pymdownx.superfences"]["custom_fences"] = [
+            {
+                # "name" specifies fences to extract.
+                "name": "*",
+                "class": "test",
+                "format": self._fence_null_format,
+            },
+        ]
+        return config["mdx_configs"]
```
improvement: Was tidying things up and moved this into the class. But it should only happen once for the entire test, instead of for each Markdown file.

inglesp reviewed (approved):
This is good to see.  I'm sure there are some more tweaks you could make, but I'd be happy for this to be merged now.
